trees/src/iavl/tree.rs

```diff
@@ -422,17 +567,141 @@ where
         }
     }
 
+    pub fn remove(
+        &mut self,
+        key: &impl AsRef<[u8]>,
+        version: u32,
+    ) -> Result<Option<Box<Node>>, Error> {
+        return tree_remove(&mut self.root, key, version, &self.node_db);
+
+        fn tree_remove<T: Database>(
+            tree: &mut Option<Box<Node>>,
+            key: &impl AsRef<[u8]>,
+            version: u32,
+            node_db: &NodeDB<T>,
+        ) -> Result<Option<Box<Node>>, Error> {
+            match tree {
+                None => return Ok(None),
+                Some(node) => match node.as_mut() {
+                    // auto deref make pain so we need to explicit call `as_mut`
+                    Node::Leaf(leaf) => {
+                        if leaf.key[..] != *key.as_ref() {
+                            return Err(Error::NodeNotExists);
+                        }
+                    }
+                    Node::Inner(inner) => {
+                        if let Some(result) = match inner.key[..].cmp(key.as_ref()) {
+                            Ordering::Less => {
+                                Some(tree_remove(&mut inner.right_node, key, version, node_db))
+                            }
+                            Ordering::Equal => None,
+                            Ordering::Greater => {
+                                Some(tree_remove(&mut inner.left_node, key, version, node_db))
+                            }
+                        } {
+                            node.rebalance(version, node_db)?;
+                            return result;
+                        }
+                    }
+                },
+            }
+
+            // If control flow fell through to here, it's because we hit the Equal case above. The
```

Comment:
Can we avoid this assumption by moving the below code into the appropriate branch above?